src/layers/layer.cpp

```diff
@@ -272,6 +275,9 @@ void Layer::forward_prop() {
 }
 
 void Layer::back_prop() {
+  if (m_frozen) {
```

Comment:
I am not a fan of this. It forces all layers "below" a frozen layer to be frozen as well. There is already support to exit backprop early if possible, so it just constrains functionality.

Reply:
Without it, bp_compute() will still call gemms and there will be GPU synchronizations.
model's backward_prop() will still call the the back_prop() of all the layers below, in case that one has multiple layers above and not all of them are frozen.

Reply:
If you look in `model::backward_prop()`, the model stops calling `layer::backward_prop()` when it has all the gradients it needs. For a simple linear network, this will only call one unnecessary GEMM.

In any case, we need the ability to backprop through frozen layers if we want to use this functionality for GANs.

Reply:
I think there are three solutions here.
1. back to the original version: put "if (l->is_frozen()) continue" into the top of the body of the for-loop in 'model::backward_prop()', and remove the "if (m_frozen) return" in 'Layer::back_prop()'.
2. initialize 'all_gradients_computed' to "!l->is_frozen()" instead of true.
3. remove "if (m_frozen) return" in 'Layer::back_prop()'

Both option 1 and option 2 do not call extra gemm or synchronization, and still allows back_prop of layers below a frozen layer.
The former is the most straight forward but skips 'do_layer_backward_prop_begin/end_cbs()'.
The latter is the opposite with some overhead.
Option 3 allows back_prop of layers below a frozen layer with extra gemm, which are in my view side effects to clean up without knowing what you have in mind for future extension.
My preference would be option 2.
Other requirement I see is that 'get_bp_output()' should zero the input matrix. For this purpose, 'Layer::add_to_error_signal()' on a frozen layer can simply zero the 'm_error_signals' instead of calling El::Axpy().

Opinions/objections?

Reply:
I would lobby for option 3. If I'm interpreting this functionality correctly, freezing weights just means that we don't need to compute gradient updates. However, an unfrozen layer below a frozen layer still needs the frozen layer's error signal to correctly compute its own gradients. We need this functionality for GANs, where we might train a generator network that outputs into a frozen discriminator network. Terminating or zeroing out the error signal of frozen layers will be incorrect for GANs, although it does save a substantial amount of computation in the Siamese networks.

Reply:
In that case, solution 2 and 3 need to be combined to avoid early termination. agreed?

Reply:
Option 2 terminates backprop once it encounters a frozen layer, which isn't the behavior we want. With option 3, early termination only happens when all required gradients have been computed.

Reply:
Probably the confusion is coming from "!l->". It looks like double el(l)s, but it is boolean not and an el. So, "!l->is_frozen()" tells if the layer 'l' is not frozen. it behaves the same with a non-frozen layer and avoids breaking the loop with a frozen layer. With option 3 alone, 'get_optimizer()' returns nullptr if a layer is frozen. This will not allow changing 'all_gradients_computed' to false.
To make freezing dynamic, we need to make sure that it is ok with callbacks. 
Currently, freezing is done before setting up callbacks, which is ok for now.